nav2_world_model/src/costmap_representation.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2018 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+#include <sstream>
+#include <memory>
+#include <vector>
+
+#include "nav2_world_model/costmap_representation.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+#include "nav2_util/angleutils.hpp"
+
+namespace nav2_world_model
+{
+
+using nav2_costmap_2d::MapLocation;
+using nav2_util::Point;
+
+CostmapRepresentation::CostmapRepresentation(
+  const std::string name,
+  rclcpp::Node::SharedPtr & node,
+  rclcpp::executor::Executor & executor,
+  rclcpp::Clock::SharedPtr & clock)
+: WorldRepresentation(name, node),
+  clock_(clock),
+  tfBuffer_(clock_),
+  tfListener_(tfBuffer_),
+  region_visualizer_(node_)
+{
+  costmap_ros_ = std::make_shared<nav2_costmap_2d::Costmap2DROS>(name_, tfBuffer_);
+  costmap_ = costmap_ros_->getCostmap();
+  executor.add_node(costmap_ros_);
+}
+
+GetCostmap::Response
+CostmapRepresentation::getCostmap(const GetCostmap::Request & /*request*/)
+{
+  GetCostmap::Response response;
+  response.map.metadata.size_x = costmap_->getSizeInCellsX();
+  response.map.metadata.size_y = costmap_->getSizeInCellsY();
+  response.map.metadata.resolution = costmap_->getResolution();
+  response.map.metadata.layer = "Master";
+  response.map.metadata.map_load_time = costmap_ros_->now();
+  response.map.metadata.update_time = costmap_ros_->now();
+
+  tf2::Quaternion quaternion;
+  // TODO(bpwilcox): Grab correct orientation information
+  quaternion.setRPY(0.0, 0.0, 0.0);  // set roll, pitch, yaw
+  response.map.metadata.origin.position.x = costmap_->getOriginX();
+  response.map.metadata.origin.position.y = costmap_->getOriginY();
+  response.map.metadata.origin.position.z = 0.0;
+  response.map.metadata.origin.orientation = tf2::toMsg(quaternion);
+
+  response.map.header.stamp = costmap_ros_->now();
+  response.map.header.frame_id = "map";
+
+  unsigned char * data = costmap_->getCharMap();
+  auto data_length = response.map.metadata.size_x * response.map.metadata.size_y;
+  response.map.data.resize(data_length);
+  response.map.data.assign(data, data + data_length);
+
+  return response;
+}
+
+ProcessRegion::Response
+CostmapRepresentation::confirmFreeSpace(const ProcessRegion::Request & request)
+{
+  ProcessRegion::Response response;
+  response.was_successful = checkIfFree(request);
+  return response;
+}
+
+ProcessRegion::Response
+CostmapRepresentation::clearArea(const ProcessRegion::Request & /*request*/)
```

Comment:
It seems to be these clearArea, confirmFreeSpace, etc methods should sit on top of the costmap implementation, not be part of it. I'm still not entirely sure what these are doing exactly given a planner or controller should look at a costmap and make the decision about the validity of the action based on the map presented. Having something in the main navigator loop that queries a configurable shape for something in the costmap or laser scan at loop rate makes sense to add an additional level of safety, but I dont think that belongs here along side t he costmaps, but maybe in the WorldModel class or as a separate utility.

Reply:
Or part of this Robot API

Reply:
I'll try to address your comment, but this can be a lenghy topic of discussion for the workgroup meeting.

We are working towards creating a world model that is independent of the underlying world representation (currently we only have costmap, but there are others). That means we need to define what any world representation should provide to the world model.

> It seems to be these clearArea, confirmFreeSpace, etc methods should sit on top of the costmap implementation, not be part of it.

That's the goal of the `CostmapRepresentation`. It's providing the inteface required by the world model to use a costmap. A costmap should be able, among other things, to confirm if some region is occupied by calling `confirmFreeSpace`.

> I'm still not entirely sure what these are doing exactly... 

These methods define the required information for the current type of world model clients used on the stack -- (global) planner and motion primitives. The DWA costmap has not been consolidated to the world model yet.

> ... given a planner or controller should look at a costmap and make the decision about the validity of the action based on the map presented.

That's an option, providing the entire costmap to the local planner, however from a design point of view, it's better to understand what exactly is the information needed by the controller. For example, it's my understanding that the dwa controller needs the costmap to check if a possible trajectory hits an object, that could be a service on the world model, instead of having each controller implement a different algorithm for checking if a trajectory goes into an object.

> Having something in the main navigator loop that queries a configurable shape for something in the costmap or laser scan at loop rate makes sense to add an additional level of safety

Yes the `confirmFreeSpace` could be used by the safetly layer, but not exclusively. Motion primitives might, possibly some type of planner. Therefore it makes sense to have it as a general service of the world model.

> but I dont think that belongs here along side t he costmaps, but maybe in the WorldModel class or as a separate utility.

It is a service of the world model and is something the costmap should provide, the `CostmapRepresentation` is just adapting the interface of the world model to the one fo the costmap, for example:

backup_primitive --(freeSpace?)--> world_model --(freeSpace?)--> world_representation

As the stack grows to support other families of controllers, planners, primitives, world representations, multiple robots (and a shared comprehesion of the world), this re-design will be necessary.